### DISCUSSION TITLE
How to use vcpkg and a manifest file with specific feature(s)/triplet?

### DISCUSSION THREAD

**Discussion by meastp** [microsoft/vcpkg#19138] [Ideas]
When asking a question please also include where you looked for an answer (so we can update the documentation if needed).

I'm investigating the new manifest feature (@strega-nil?), and have the following use case:

My example project `myproject` has three features: `client` `client-ext` (depends on client) and `server` (depends on both client and client-ext). Default should be `server` which builds/installs all of the dependencies. However, for some configurations (custom triplets) like x86-windows-v141 only the client-ext feature (which means all dependencies for client and client-ext) should be built/installed.

I could not understand from the documentation how the `vcpkg install` with a `vcpkg.json` file handles this scenario?

1. **How to select which feature(s) to install?**
2. **How to choose which triplet (or triplets) to install?**

**Comment by strega-nil** [comment]
I would say probably that the existing hack is, unfortunately, the best thing to do:

```json
{
  "name": "myproject",
  "version-string": "...",
  "default-features": [ "default-features" ],
  "features": [
    {
      "name": "default-features",
      "dependencies": [
        {
          "name": "myproject",
          "features": [ "client", "client-ext" ],
          "platform": "x86 & windows & v141"
        },
        {
          "name": "myproject",
          "features": [ "server" ],
          "platform": "!(x86 & windows & v141)"
        }
      ]
    },
    {
      "name": "client",
      "description": "..."
    },
    {
      "name": "client-ext",
      "description": "..."
    },
    {
      "name": "server",
      "description": "..."
    }
  ]
}
```

this is how we solve it in control files, anyways. I know this sucks, but we haven't yet done the engineering work to fix it.

**Comment by meastp** [comment]
@strega-nil thanks - as long as it works, it's okay as a temporary solution, no worries :)

However, my second question was/is:

When I'm in a terminal (powershell) at `C:\Repos\myproject` which contains this `vcpkg.json` file (containing your example above): **How do I invoke `vcpkg.exe` with `vcpkg.json` to install myproject-dependencies for a specific and/or all triplets I need?**

**Comment by strega-nil** [comment]
`vcpkg install --feature-flags=manifests` will work, or if you want to install into a specific directory `vcpkg install --feature-flags=manifests --x-install-root=<install directory>`.

The triplet can be defined via the `--triplet <triplet>` flag, if you want specific features, you can use `--x-feature=blah --x-feature=foo`, and `--x-no-default-features` if you don't want the default features.

[ed. note: sorry that all of the flags are passed differently, I hope to work on this soon]

**Comment by meastp** [comment]
Thanks! I'm having problems with the VCPKG_PLATFORM_TOOLSET, though:
```
$vcpkg install --triplet x64-windows-v141 --feature-flags=manifests
Error: Unrecognized identifer name v141. Add to override list in triplet file.
Error: Unrecognized identifer name v141. Add to override list in triplet file.
Detecting compiler hash for triplet x64-windows-v141...
(...)
```
When I change from `"platform": "x64 & windows & v141"` to `"platform": "x64 & windows"` no error messages appear. I need to support VCPKG_PLATFORM_TOOLSET in the platform specification. Is this possible? How can I patch the vcpkg project locally to add support for this?

**Comment by BillyONeal** [comment]
You can't; adding new things to platform-expression will require changes to `vcpkg.exe` itself. See for example https://github.com/microsoft/vcpkg/blob/master/toolsrc/src/vcpkg/platform-expression.cpp#L37

**Comment by meastp** [comment]
@BillyONeal That's inconvenient for me as I differentiate on the VCPKG_PLATFORM_TOOLSET. Could this be added to platform-expressions, please? :) An alternative is to be able to specify the whole triplet.
